### PR TITLE
Add support for shared XCCDF for Debian based OSes

### PR DESCRIPTION
**Pull request by pthierry38**
Hi,

As discussed some time ago with Martin...
I've updated the Debian-based OSes to support shared XCCDF instead of maintaining separately 3 XCCDF content with lots of duplication.
This took me some work, which is one of the two root cause of the size of this PR :-/
This PR:
- add Debian9 (second root cause)
- update the shared/xccdf part to support debian8, debian9, ubuntu14.04 ubuntu16.04
- update the shared oval consequently and add some deb-specific oval into Debian/* and Ubuntu/*
   * there is some diff mostly in pathnames, it may be possible to patch the sared Oval instead of overloading it, but this would make the shared oval really hard to read
- add some Debian/Ubuntu specific rules that was in each XCCDF guides before in the shared xccdf files, using the prodtype= tag, to avoid any impact on other OSes.

There is still some missing OVAL/Bench ID at build time for Deb* and Ubu*, but which don't impact the production. I will make some *small* patch to correct this. Maybe it would be better to pass this PR as WIP.

The build has been tested on Debian 9.0 (stretch), with one little other patch due to Issue #2130. This little patch is of course not a part of this PR.

### CODE REVIEW

**Comment by redhatrises** [comment]
@pthierry38 can you remove adding `Debian 9` from this PR and have a separate PR for it? It makes it incredibly difficult to focus reviewing the updates to adding Debian to the shared portion of the XCCDF which will need a bit of scrutiny.

**Comment by pthierry38** [comment]
Ok il will extract this commit from this PR

**Comment by mpreisler** [comment]
@pthierry38 please rebase and get rid of the debian9 commits, we don't want a revert commit in a PR.

**Comment by pthierry38** [comment]
revert commit and deb9 content commit deleted via rebase

File: Debian/8/templates/static/oval/audit_rules_dac_modification_chmod.xml [hunk 1]
```diff
@@ -0,0 +1,77 @@
+<def-group>
```
Comment (mpreisler) [inline Debian/8/templates/static/oval/audit_rules_dac_modification_chmod.xml]: we have templates for these now

**Comment by mpreisler** [comment]
@pthierry38 are the static OVALs all written specifically for Ubuntu and Deb or can we share them with other distros?

**Comment by pthierry38** [comment]
The ones i put in debian / Ubuntu dir can't be shared easily (differences in binary path, etc.). I uszd as much shared ival as i could.

File: shared/xccdf/system/permissions/files.xml [hunk 2]
```diff
@@ -42,6 +42,17 @@ critical for system security.</rationale>
 <ref nist="AC-6" disa="" pcidss="Req-8.7.c" cjis="5.5.2.2" />
 </Rule>
 
+<Rule id="groupowner_shadow_file" severity="medium" prodtype="debian8,ubuntu1404">
```
Comment (redhatrises) [inline shared/xccdf/system/permissions/files.xml]: Nack. This rule should be merged into the existing rule, and `prodtype` added to `<description />` and `<ocil />` elements of the existing rule.

File: shared/xccdf/system/permissions/files.xml [hunk 3]
```diff
@@ -111,6 +122,18 @@ is critical for system security.</rationale>
 <ref nist="AC-6" disa=""/>
 </Rule>
 
+<Rule id="file_groupowner_etc_gshadow" severity="medium" prodtype="debian8,ubuntu1404">
```
Comment (redhatrises) [inline shared/xccdf/system/permissions/files.xml]: Nack. This rule should be merged into the existing rule, and `prodtype` added to `<description />` and `<ocil />` elements of the existing rule.

File: shared/xccdf/system/software/disk_partitioning.xml [hunk 4]
```diff
@@ -104,10 +104,10 @@ of space.
 <ident prodtype="rhel7" cce="26971-2" />
 <oval id="partition_for_var_log_audit" />
 <ref prodtype="rhel7" stigid="021330" />
-<ref nist="AU-4,AU-9,SC-32(1)" disa="366" cis="1.1.8" ossrg="SRG-OS-000480-GPOS-00227" />
+<ref anssi="NT28(R12)" nist="AU-4,AU-9,SC-32(1)" disa="366" cis="1.1.8" ossrg="SRG-OS-000480-GPOS-00227" />
```
Comment (redhatrises) [inline shared/xccdf/system/software/disk_partitioning.xml]: Does CIS 1.1.8 map to the Debian profile or this to RHEL?

File: shared/xccdf/services/obsolete.xml [hunk 5]
```diff
@@ -157,6 +157,30 @@ Removing the <tt>telnet-server</tt> package decreases the risk of the telnet ser
...
+<br />
+The telnet service provides an unencrypted remote access service which does not provide for the 
+confidentiality and integrity of user passwords or the remote session. If a privileged user were
+to login using this service, the privileged user password could be compromised.
+<br />
+Removing the <tt>telnet-server</tt> package decreases the risk of the telnet service's accidental 
+(or intentional) activation.
+</rationale>
+<oval id="package_telnetd_removed" />
+<ref anssi="NT007(R03)" nist="AC-17(8),CM-7(a)" disa="381" ossrg="SRG-OS-000095-GPOS-00049" cis="2.1.1" />
```
Comment (redhatrises) [inline shared/xccdf/services/obsolete.xml]: Does this CIS 2.1.1 map to RHEL or Debian?

File: shared/xccdf/services/obsolete.xml [hunk 6]
```diff
@@ -172,6 +196,20 @@ Enterprise Linux.</rationale>
 <ref cis="2.1.2" cui="3.1.13" />
 </Rule>
 
+<!-- not using yum macro but apt. shared oval file should be overloaded by local debian/ubuntu csv -->
+<Rule id="package_telnet_removed" severity="low" prodtype="debian8,ubuntu1404">
```
Comment (redhatrises) [inline shared/xccdf/services/obsolete.xml]: If this rule already exists, it should be collapsed into an existing rule and the appropriate elements updated with `prodtype`

File: shared/xccdf/services/obsolete.xml [hunk 7]
```diff
@@ -180,11 +218,13 @@ Enterprise Linux.</rationale>
 allow cleartext remote access and have an insecure trust
 model.</description>
 
-<Rule id="package_rsh-server_removed" severity="high" prodtype="rhel7">
+<Rule id="package_rsh-server_removed" severity="high" prodtype="rhel7,debian8,ubuntu1404">
 <title>Uninstall rsh-server Package</title>
 <description>The <tt>rsh-server</tt> package can be uninstalled with
```
Comment (redhatrises) [inline shared/xccdf/services/obsolete.xml]: Please undo the description changes here. Use 2 separate descriptions with prodtype that makes the target os.

File: shared/templates/static/oval/installed_OS_is_debian9.xml [hunk 8]
```diff
@@ -14,6 +14,8 @@
     <criteria operator="AND">
       <extend_definition comment="Installed OS is part of the Unix family"
       definition_ref="installed_OS_is_part_of_Unix_family" />
+      <extend_definition comment="Installed OS is Debian"
```
Comment (redhatrises) [inline shared/templates/static/oval/installed_OS_is_debian9.xml]: This seems like a redundant check to add.

File: shared/templates/static/oval/installed_OS_is_debian8.xml [hunk 9]
```diff
@@ -14,6 +14,8 @@
     <criteria operator="AND">
       <extend_definition comment="Installed OS is part of the Unix family"
       definition_ref="installed_OS_is_part_of_Unix_family" />
+      <extend_definition comment="Installed OS is Debian"
```
Comment (redhatrises) [inline shared/templates/static/oval/installed_OS_is_debian8.xml]: This seems like a redundant check to add.

File: shared/templates/static/oval/partition_for_tmp.xml [hunk 10]
```diff
@@ -3,7 +3,7 @@
     <metadata>
       <title>Ensure /tmp Located On Separate Partition</title>
       <affected family="unix">
-        <platform>multi_platform_oval</platform>
+        <platform>multi_platform_all</platform>
```
Comment (redhatrises) [inline shared/templates/static/oval/partition_for_tmp.xml]: This should be `<platform>multi_platform_oval</platform>`

File: shared/templates/static/oval/partition_for_var.xml [hunk 11]
```diff
@@ -3,7 +3,7 @@
     <metadata>
       <title>Ensure /var Located On Separate Partition</title>
       <affected family="unix">
-        <platform>multi_platform_oval</platform>
+        <platform>multi_platform_all</platform>
```
Comment (redhatrises) [inline shared/templates/static/oval/partition_for_var.xml]: This should be `<platform>multi_platform_oval</platform>`

File: shared/templates/static/oval/partition_for_var_log.xml [hunk 12]
```diff
@@ -3,7 +3,7 @@
     <metadata>
       <title>Ensure /var/log Located On Separate Partition</title>
       <affected family="unix">
-        <platform>multi_platform_oval</platform>
+        <platform>multi_platform_all</platform>
```
Comment (redhatrises) [inline shared/templates/static/oval/partition_for_var_log.xml]: This should be `<platform>multi_platform_oval</platform>`

File: shared/templates/static/oval/partition_for_var_log_audit.xml [hunk 13]
```diff
@@ -4,6 +4,8 @@
       <title>Ensure /var/log/audit Located On Separate Partition</title>
       <affected family="unix">
         <platform>multi_platform_oval</platform>
+        <platform>multi_platform_debian</platform>
```
Comment (redhatrises) [inline shared/templates/static/oval/partition_for_var_log_audit.xml]: Don't think that these are needed with `<platform>multi_platform_oval</platform>`

File: shared/templates/static/oval/partition_for_var_log_audit.xml [hunk 14]
```diff
@@ -4,6 +4,8 @@
...
+        <platform>multi_platform_ubuntu</platform>
```
Comment (redhatrises) [inline shared/templates/static/oval/partition_for_var_log_audit.xml]: Don't think that these are needed with `<platform>multi_platform_oval</platform>`

File: shared/xccdf/system/auditing.xml [hunk 15]
```diff
@@ -100,7 +100,30 @@ the process, which in this case, is <tt>exe="/usr/sbin/httpd"</tt>.
 </li></ul>
 </description>
 
-<Rule id="service_auditd_enabled" severity="high" prodtype="rhel7">
+<Rule id="package_auditd_installed" severity="high" prodtype="debian8,debian9,ubuntu1404">
+<title>Install auditd Service</title>
```
Comment (redhatrises) [inline shared/xccdf/system/auditing.xml]: This should be `Package` and not `Service`

File: shared/xccdf/system/auditing.xml [hunk 16]
```diff
@@ -100,7 +100,30 @@ the process, which in this case, is <tt>exe="/usr/sbin/httpd"</tt>.
...
+Ensuring the <tt>auditd</tt> service is active ensures audit records
+generated by the kernel are appropriately recorded.
+<br /><br />
+Additionally, a properly configured audit subsystem ensures that actions of 
+individual system users can be uniquely traced to those users so they
+can be held accountable for their actions.
+</rationale>
+<platform idref="cpe:/a:machine" />
+<oval id="package_auditd_installed" />
+<ref anssi="NT28(R50)" nist="AU-3,AC-17(1),AU-1(b),AU-10,AU-12(a),AU-12(c),AU-14(1),IR-5" disa="126,131" ossrg="SRG-OS-000038-GPOS-00016,SRG-OS-000039-GPOS-00017,SRG-OS-000042-GPOS-00021,SRG-OS-000254-GPOS-00095,SRG-OS-000255-GPOS-00096" pcidss="Req-10" cis="5.2.2" cjis="5.4.1.1" cui="3.3.1,3.3.2,3.3.6" />
```
Comment (redhatrises) [inline shared/xccdf/system/auditing.xml]: Do all these references line up to match that the `auditd` package should be installed?

File: shared/xccdf/system/logging.xml [hunk 17]
```diff
@@ -12,12 +12,14 @@ In Red Hat Enterprise Linux 7, rsyslog has replaced ksyslogd as the
 syslog daemon of choice, and it includes some additional security features
 such as reliable, connection-oriented (i.e. TCP) transmission of logs, the
 option to log to database formats, and the encryption of log data en route to
-a central logging server.
+a central logging server.<br />
+In Debian and Ubuntu, rsyslog is on of the two main logservers, with syslog-ng,
```
Comment (redhatrises) [inline shared/xccdf/system/logging.xml]: This shouldn't be added here. It should be in a separate description with `prodtype`

File: shared/xccdf/system/permissions/files.xml [hunk 18]
```diff
@@ -179,9 +179,29 @@ is critical for system security.</rationale>
 <oval id="file_permissions_etc_passwd" />
 <ref nist="AC-6" disa="" pcidss="Req-8.7.c" cjis="5.5.2.2" />
 </Rule>
+
+<Rule id="file_permissions_systemmap" prodtype="debian8,debian9,ubuntu1404">
```
Comment (redhatrises) [inline shared/xccdf/system/permissions/files.xml]: Please remove line indentation from this new rule.

File: shared/xccdf/system/permissions/files.xml [hunk 19]
```diff
@@ -179,9 +179,29 @@ is critical for system security.</rationale>
...
+	can give some hints to generate local exploitation.
+</rationale>
+<oval id="file_permissions_systemmap" />
+<ref anssi="NT28(R13)" />
+</Rule>
+
 </Group>
 
 
+
```
Comment (redhatrises) [inline shared/xccdf/system/permissions/files.xml]: Why an extra whitespace?

File: shared/templates/static/oval/installed_OS_is_debian9.xml [hunk 20]
```diff
@@ -1,27 +1,31 @@
 <def-group>
-  <definition class="inventory"
-  id="installed_OS_is_debian9" version="2">
+  <definition class="inventory" id="installed_OS_is_debian9" version="1">
     <metadata>
       <title>Debian 9</title>
       <affected family="unix">
-        <platform>Debian 9</platform>
+        <platform>multi_platform_all</platform>
```
Comment (redhatrises) [inline shared/templates/static/oval/installed_OS_is_debian9.xml]: Shouldn't this be `<platform>multi_platform_oval</platform>`?

File: shared/templates/static/oval/installed_OS_is_debian9.xml [hunk 21]
```diff
@@ -1,27 +1,31 @@
...
     </metadata>
     <criteria operator="AND">
-      <extend_definition comment="Installed OS is part of the Unix family"
-      definition_ref="installed_OS_is_part_of_Unix_family" />
-      <extend_definition comment="Installed OS is Debian"
-      definition_ref="installed_OS_is_debian" />
-      <criterion comment="Debian9 is installed"
-      test_ref="test_debian9" />
+      <extend_definition comment="Installed OS is part of the Unix family" definition_ref="installed_OS_is_part_of_Unix_family" />
+      <criterion comment="Debian version file exists" test_ref="test_debian9_file" />
```
Comment (redhatrises) [inline shared/templates/static/oval/installed_OS_is_debian9.xml]: Is this check needed? Does it not fail with `test_debian9` check?

File: shared/templates/static/oval/partition_for_var.xml [hunk 22]
```diff
@@ -3,7 +3,9 @@
     <metadata>
       <title>Ensure /var Located On Separate Partition</title>
       <affected family="unix">
-        <platform>multi_platform_all</platform>
+        <platform>multi_platform_rhel</platform>
+        <platform>multi_platform_debian</platform>
+        <platform>multi_platform_ubuntu</platform>
```
Comment (redhatrises) [inline shared/templates/static/oval/partition_for_var.xml]: I believe that this should only be `<platform>multi_platform_oval</platform>`

File: shared/templates/static/oval/partition_for_var_log.xml [hunk 23]
```diff
@@ -3,7 +3,9 @@
     <metadata>
       <title>Ensure /var/log Located On Separate Partition</title>
       <affected family="unix">
-        <platform>multi_platform_all</platform>
+        <platform>multi_platform_rhel</platform>
+        <platform>multi_platform_debian</platform>
+        <platform>multi_platform_ubuntu</platform>
```
Comment (redhatrises) [inline shared/templates/static/oval/partition_for_var_log.xml]: I believe that this should only be `<platform>multi_platform_oval</platform>`

File: shared/xccdf/system/permissions/execution.xml [hunk 24]
```diff
@@ -242,7 +242,6 @@ address information.</rationale>
 <description><sysctl-desc-macro sysctl="kernel.sysrq" value="0" /></description>
 <ocil><sysctl-check-macro sysctl="kernel.sysrq" value="0" /></ocil>
 <rationale>Unprivileged access to the system lifecycle through the sysfs may impact disponibility and current security state of the target.</rationale>
-<platform idref="cpe:/a:machine" />
```
Comment (redhatrises) [inline shared/xccdf/system/permissions/execution.xml]: Why was this removed?

File: shared/xccdf/system/software/integrity.xml [hunk 25]
```diff
@@ -52,11 +52,14 @@ software update.  AIDE is highly configurable, with further configuration
 information located in <tt>/usr/share/doc/aide-<i>VERSION</i></tt>.
 </description>
 
-<Rule id="package_aide_installed" severity="medium" prodtype="rhel7">
+<Rule id="package_aide_installed" severity="medium" prodtype="rhel7,debian9">
 <title>Install AIDE</title>
 <description>
-Install the AIDE package with the command:
+Install the AIDE package. It can be done using with the command:
```
Comment (redhatrises) [inline shared/xccdf/system/software/integrity.xml]: This change should not be made. Please use description with `prodtype`.

File: shared/xccdf/system/software/integrity.xml [hunk 26]
```diff
@@ -52,11 +52,14 @@ software update.  AIDE is highly configurable, with further configuration
...
 <pre>$ sudo yum install aide</pre>
+on Red-Hat or the command:
+<pre>$ sudo apt install aide</pre>
+on Debian.
```
Comment (redhatrises) [inline shared/xccdf/system/software/integrity.xml]: This change should not be made. Please use description with `prodtype`.

File: shared/xccdf/services/ntp.xml [hunk 27]
```diff
@@ -69,6 +69,41 @@ information on the capabilities and configuration of each of the NTP daemons.
 <value selector="disa">17</value>
 </Value>
 
+<Rule id="service_chrony_or_ntp_enabled" severity="medium" prodtype="debian8,debian9">
```
Comment (redhatrises) [inline shared/xccdf/services/ntp.xml]: Why was this rule added?

Comment (redhatrises) [inline shared/xccdf/services/ntp.xml]: Nevermind. Ignore comment.

File: shared/xccdf/services/ntp.xml [hunk 28]
```diff
@@ -69,6 +69,41 @@ information on the capabilities and configuration of each of the NTP daemons.
...
+<title>Enable the NTP Daemon</title>
+<description>
+<systemd-service-macro enable="true" service="chrony" />
+Note: The <tt>chrony</tt> daemon is enabled by default.
+<br /><br />
+<systemd-service-macro enable="true" service="ntp" />
+Note: The <tt>ntp</tt> daemon is not enabled by default. Though as mentioned
+in the previous sections in certain environments the <tt>ntp</tt> daemon might
+be preferred to be used rather than the <tt>chrony</tt> one. Refer to:
+<weblink-macro link="https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/System_Administrators_Guide/ch-Configuring_NTP_Using_the_chrony_Suite.html"/>
```
Comment (redhatrises) [inline shared/xccdf/services/ntp.xml]: Why are you linking to Red Hat for the Debian product?

**Comment by redhatrises** [comment]
@pthierry38 in the future, please make smaller and more PR requests rather than one massive one. It makes reviews as well as updates much easier for everyone involved.

**Comment by pthierry38** [comment]
@redhatrises I've taken into account your review, but i don't understand why the rhel7 fails with:
```
[rhel7-tables] generating HTML table for cis references
runtime error: file /home/jenkins/workspace/scap-security-guide-pull-requests/label/node-el6-openscap-new/shared/transforms/shared_xccdf2table-byref.xslt line 195 element copy
Attribute nodes must be added before any child nodes to an element.
```
I've reviews the few patches of this morning modifying the way shared xccdf is handle but i don't see where it fails.
The patches use prodtype= in various elements (definition, rationale and ocil) to separate RH specific and Deb specific.
Do you know where is the error ?

**Comment by redhatrises** [comment]
@pthierry38 I apologize. `multi_platform_oval` is not correct. It is actually `multi_platform_all`. Not sure what I was thinking.

**Comment by pthierry38** [comment]
I've updated the platform target but the error is still here... I will try to find out why but i haven't got a clue about where it comes from

**Comment by pthierry38** [comment]
Found! the usage of prodtype="" in rationale elements is not handled properly, which leads to duplicate rationale elements in the ds file, making the xslt script fails.
I'm opening an issue and try to find out where it come from

**Comment by mpreisler** [comment]
@pthierry38 I am really sorry but this PR got too far from upstream because we weren't able to review it in time. Could you please take a look and rebase it? It's impossible to review in this state.

**Comment by redhatrises** [comment]
@mpreisler @pthierry38 might need the new license to continue working on this.

**Comment by mpreisler** [comment]
Closed for inactivity.

**Pull request closed without merging**